### PR TITLE
chore: [IOAPPFD0-10] Startup saga, onboarding refactoring

### PR DESCRIPTION
**Pull request by Vangaorth**
# ⚠️ This pr depends on #4147 ⚠️
Please review #4147 first in order to avoid duplicate review work.

## Short description
This PR refactors the onboarding branching on the startup saga, in order to have a cleaner code

## List of changes proposed in this pull request
This PR aims to simplify the startup saga by re-arranging the onboarding branch condition.

The current state of the code is shown in the following diagram

The simplification process focuses on removing the direct flow on the right side of the diagram, which skips all the intermediate calls.

That condition happens when these three conditions are all met: there is a previous session token, there is a local stored pin and the session was refreshed. Specifically, we focus on the last of the three conditions (i.e., the session was refreshed).
If it is met, then the `previousSessionToken` must be different from the `sessionToken`. This happens only if the `previousSessionToken` value is `falsy` (otherwise, the two tokens will have the same value and the condition is not met). Now, having a `falsy` `previousSessionToken` means that the branch containing the condition is not even executed, since the first of the initial three conditions (i.e., there is a previous session token) is not met anymore.

With the consideration above, we deduct that the flow is never executed and can be safely removed, leading to the following simplified diagram

From this, we can see that there are some common steps, executed on both branches. Our proposal is to completely remove the branching and execute the steps using guards that maps the branching condition, formally:

- current execution has both session and pin;
- current execution has no session or no pin (which is the logical negation of the previous)

Other conditions are mapped adding additional guards to either one or the other.

This last diagram is the final code refactoring that is being proposed with this PR:

Note that we need to compute the `isFirstOnboarding` parameter before `call(askServicesPreferencesModeOptin(isFirstOnboarding: boolean))`.

Also note that, the flow condition `(has no session OR has no pin) AND isFirstOnboarding` can be further simplified to just check `isFirstOnboarding`, since if the user does not have a session, the code flow is interrupted long before reaching this point and the same goes for the not having a pin.

## How to test
In order to test this PR you have to go through the onboarding process so start an instance of the io-app-dev-server with the `firstOnboarding` set to `true` and make sure to reinstall the application (i.e., no previously saved local data on the device/simulator). In this case, you should see all the onboarding screens (namely, ToS, Mixpanel OptIn, Pin Configuration, Fingerprint (on device), email ACK, push notifications OptIn, service configuration and onboarding completion).

After this, you can test again, by restarting the io-app-dev-server with the `firstOnboarding` flag set to `false` (do not reinstall the application). In this case, you should only see the PIN screen and the push notification OptIn.

Also check all the logout/reset features during the PIN verification and during the onboarding.

### CODE REVIEW

**Review by francescopersico — CHANGES_REQUESTED** [review]

File: ts/components/ui/BlockButtons.tsx [hunk 1]
```diff
@@ -45,6 +49,16 @@ export interface SingleButton extends CommonProps {
   type: "SingleButton";
 }
 
+/**
+ * |  left  |
+ * |  right |
+ */
+export interface TwoButtonsVertical extends CommonProps {
```
[the remark this answers is not in the document]
Comment (Vangaorth) [reply ts/components/ui/BlockButtons.tsx]: Bad code from a previous PR. Removed

File: ts/sagas/startup.ts [hunk 2]
```diff
@@ -310,69 +310,62 @@ export function* initializeApplicationSaga(): Generator<
   // Start watching for requests of abort the onboarding
   const watchAbortOnboardingSagaTask = yield* fork(watchAbortOnboardingSaga);
 
-  if (!previousSessionToken || O.isNone(maybeStoredPin)) {
-    // The user wasn't logged in when the application started or, for some
-    // reason, he was logged in but there is no unlock code set, thus we need
-    // to pass through the onboarding process.
+  // This block has been refactored to avoid a code duplication on an if-else branching.
```
[the remark this answers is not in the document]
Comment (Vangaorth) [reply ts/sagas/startup.ts]: Comments updated

File: ts/sagas/startup.ts [hunk 3]
```diff
@@ -316,69 +316,57 @@ export function* initializeApplicationSaga(): Generator<
...
+  } else {
+    // TODO If the session was not valid, the code would have stopped before
+    // reaching this point. Consider refactoring even more by removing the check
+    // on the session (IOAPPCIT-10)
     storedPin = yield* call(checkConfiguredPinSaga);
 
     yield* call(checkAcknowledgedFingerprintSaga);
 
     yield* call(checkAcknowledgedEmailSaga, userProfile);
+  }
```
Comment (CrisTofani) [inline ts/sagas/startup.ts:354]: Could this block of code be moved inside the `if` clause at line [321](https://github.com/pagopa/io-app/pull/4168/files#diff-8a5b2f3967d681b976fe673762bd1061f5b430130c880c1195b76af06362cf31R321)?

**Review by CrisTofani — CHANGES_REQUESTED** [review]

**Review by CrisTofani — APPROVED** [review]
LGTM @shadowsheep1 @francescopersico  waiting your approval to merge

**Review by shadowsheep1 — APPROVED** [review]

File: ts/sagas/startup.ts [hunk 4]
```diff
@@ -316,69 +316,57 @@ export function* initializeApplicationSaga(): Generator<
...
-    yield* call(checkAcceptedTosSaga, userProfile);
+  // Ask to accept ToS if there is a new available version
+  yield* call(checkAcceptedTosSaga, userProfile);
 
-    // check if the user expressed preference about mixpanel, if not ask for it
-    yield* call(askMixpanelOptIn);
+  // check if the user expressed preference about mixpanel, if not ask for it
+  yield* call(askMixpanelOptIn);
 
+  if (hasPreviousSessionAndPin) {
```
[the remark this answers is not in the document]
Comment (Vangaorth) [reply ts/sagas/startup.ts:341]: That's due to the scope of this PR, which was to simplify a branching condition with duplicate code. Removing the check for the session may be possible but it is out of the PR scope (here's the tracking https://pagopa.atlassian.net/browse/IOAPPCIT-10)

**Review by francescopersico — APPROVED** [review]

**Pull request merged**